File: sdks/python/finault/otel_exporter.py
"""
Finault OpenTelemetry Exporter (Python)
═════════════════════════════════════════════════════════════════════════════════

OTel span exporter that emits spans with Finault cost, margin, and seal attributes
Compatible with opentelemetry-api standard
"""

import json
import time
import threading
import requests
from typing import List, Optional, Dict, Any
from opentelemetry.sdk.trace import ReadableSpan, SpanExporter
from opentelemetry.sdk.trace.export import SpanExportResult


class FinaultOtelExporter(SpanExporter):
    """
    Finault OpenTelemetry Span Exporter

    Exports spans with Finault-specific attributes (cost, margin, seal data)
    to the Finault telemetry API.
    """

    def __init__(
        self,
        org_id: str,
        api_key: str,
        endpoint: str = "https://gateway.finault.ai/api/telemetry",
        batch_size: int = 100,
        flush_interval: float = 5.0
    ):
        """
        Initialize Finault OTel exporter

        Args:
            org_id: Organization ID
            api_key: Finault API key
            endpoint: Telemetry API endpoint
            batch_size: Number of spans to batch before flushing
            flush_interval: Seconds between auto-flushes
        """
        self.org_id = org_id
        self.api_key = api_key
        self.endpoint = endpoint
        self.batch_size = batch_size
        self.flush_interval = flush_interval

        self.buffer: List[ReadableSpan] = []
        self.buffer_lock = threading.Lock()
        self.flush_timer: Optional[threading.Timer] = None

    def export(self, spans: List[ReadableSpan]) -> SpanExportResult:
        """Export spans"""
        try:
            with self.buffer_lock:
                self.buffer.extend(spans)

                # Auto-flush if buffer exceeds batch size
                if len(self.buffer) >= self.batch_size:
                    self._flush()
                else:
                    # Schedule flush if not already scheduled
                    self._schedule_flush()

            return SpanExportResult.SUCCESS
        except Exception as err:
            print(f"Failed to export spans: {err}")
            return SpanExportResult.FAILED_NOT_RETRYABLE

    def shutdown(self) -> None:
        """Shutdown exporter"""
        if self.flush_timer:
            self.flush_timer.cancel()

        with self.buffer_lock:
            self._flush()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        """Force flush buffered spans"""
        if self.flush_timer:
            self.flush_timer.cancel()
            self.flush_timer = None

        with self.buffer_lock:
            self._flush()

        return True

    def _schedule_flush(self) -> None:
        """Schedule automatic flush"""
        if not self.flush_timer:
            self.flush_timer = threading.Timer(
                self.flush_interval,
                self._flush_thread_safe
            )
            self.flush_timer.daemon = True
            self.flush_timer.start()

    def _flush_thread_safe(self) -> None:
        """Thread-safe flush"""
        with self.buffer_lock:
            self._flush()

    def _flush(self) -> None:
        """Flush buffered spans"""
        if not self.buffer:
            return

        spans_to_export = self.buffer[:self.batch_size]
        self.buffer = self.buffer[self.batch_size:]

        try:
            telemetry_events = [
                self._span_to_telemetry_event(span)
                for span in spans_to_export
            ]

            payload = {
                "events": telemetry_events,
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            }

            headers = {
                "Content-Type": "application/json",
                "X-Finault-Org-ID": self.org_id,
                "X-Finault-API-Key": self.api_key
            }

            requests.post(self.endpoint, json=payload, headers=headers, timeout=10)
            print(f"Exported {len(telemetry_events)} telemetry events")

        except Exception as err:
            print(f"Failed to flush spans: {err}")
            # Re-add spans to buffer for retry
            self.buffer = spans_to_export + self.buffer

        # Schedule next flush if buffer still has items
        if self.buffer:
            self._schedule_flush()

    def _span_to_telemetry_event(self, span: ReadableSpan) -> Dict[str, Any]:
        """Convert span to Finault telemetry event"""
        attributes = span.attributes or {}

        # Convert timestamps from nanoseconds to milliseconds
        start_time_ms = span.start_time / 1e6
        end_time_ms = span.end_time / 1e6
        duration_ms = (span.end_time - span.start_time) / 1e6

        return {
            "trace_id": span.context.trace_id,
            "span_id": span.context.span_id,
            "parent_span_id": span.parent.span_id if span.parent else None,
            "name": span.name,
            "start_time": start_time_ms,
            "end_time": end_time_ms,
            "duration_ms": duration_ms,

            # Finault-specific attributes
            "finault": {
                "cost_usd": attributes.get("finault.cost"),
                "margin_usd": attributes.get("finault.margin"),
                "margin_percent": attributes.get("finault.margin_percent"),
                "seal_id": attributes.get("finault.seal_id"),
                "seal_url": attributes.get("finault.seal_url"),
                "model": attributes.get("finault.model"),
                "provider": attributes.get("finault.provider"),
                "tokens_in": attributes.get("finault.tokens_in"),
                "tokens_out": attributes.get("finault.tokens_out"),
                "cache_hit": attributes.get("finault.cache_hit"),
                "cost_method": attributes.get("finault.cost_method")
            },

            # Standard attributes
            "attributes": {
                "http.method": attributes.get("http.method"),
                "http.url": attributes.get("http.url"),
                "http.status_code": attributes.get("http.status_code"),
                "http.host": attributes.get("http.host"),
                "http.scheme": attributes.get("http.scheme"),
                "component": attributes.get("component")
            },

            # Status
            "status": {
                "code": span.status.status_code.name if span.status else None,
                "message": span.status.description if span.status else None
            },

            # Events
            "events": [
                {
                    "name": event.name,
                    "timestamp": event.timestamp / 1e6,
                    "attributes": event.attributes or {}
                }
                for event in (span.events or [])
            ]
        }


def init_finault_otel(
    org_id: str,
    api_key: str,
    endpoint: Optional[str] = None,
    auto_instrument: bool = False
) -> FinaultOtelExporter:
    """
    Initialize Finault OpenTelemetry exporter

    Args:
        org_id: Organization ID
        api_key: Finault API key
        endpoint: Optional custom endpoint
        auto_instrument: Whether to auto-instrument (future use)

    Returns:
        FinaultOtelExporter instance
    """
    return FinaultOtelExporter(
        org_id=org_id,
        api_key=api_key,
        endpoint=endpoint or "https://gateway.finault.ai/api/telemetry"
    )


__all__ = ["FinaultOtelExporter", "init_finault_otel"]

File: sdks/node/src/otel-exporter.ts
/**
 * Finault OpenTelemetry Exporter (Node.js)
 * ═══════════════════════════════════════════════════════════════════════════════
 *
 * OTel span exporter that emits spans with Finault cost, margin, and seal attributes
 * Compatible with @opentelemetry/api standard
 */

import {
  SpanExporter,
  ReadableSpan,
  ExportResult
} from '@opentelemetry/sdk-trace-node';

/**
 * Finault OpenTelemetry Exporter
 */
export class FinaultOtelExporter implements SpanExporter {
  private orgId: string;
  private apiKey: string;
  private endpoint: string;
  private batchSize: number;
  private flushInterval: number;
  private buffer: ReadableSpan[] = [];
  private flushTimer: NodeJS.Timeout | null = null;

  constructor(options: {
    orgId: string;
    apiKey: string;
    endpoint?: string;
    batchSize?: number;
    flushInterval?: number;
  }) {
    this.orgId = options.orgId;
    this.apiKey = options.apiKey;
    this.endpoint = options.endpoint || 'https://gateway.finault.ai/api/telemetry';
    this.batchSize = options.batchSize || 100;
    this.flushInterval = options.flushInterval || 5000; // 5 seconds
  }

  /**
   * Export spans
   */
  async export(
    spans: ReadableSpan[]
  ): Promise<ExportResult> {
    try {
      this.buffer.push(...spans);

      // Auto-flush if buffer exceeds batch size
      if (this.buffer.length >= this.batchSize) {
        await this.flush();
      } else {
        // Schedule flush if not already scheduled
        if (!this.flushTimer) {
          this.flushTimer = setTimeout(() => this.flush(), this.flushInterval);
        }
      }

      return ExportResult.SUCCESS;
    } catch (err) {
      console.error('Failed to export spans:', err);
      return ExportResult.FAILED_NOT_RETRYABLE;
    }
  }

  /**
   * Shutdown exporter
   */
  async shutdown(): Promise<void> {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
    }
    await this.flush();
  }

  /**
   * Force flush
   */
  async forceFlush(timeoutMs?: number): Promise<void> {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    await this.flush();
  }

  /**
   * Flush buffered spans
   */
  private async flush(): Promise<void> {
    if (this.buffer.length === 0) {
      return;
    }

    const spansToExport = this.buffer.splice(0, this.batchSize);

    try {
      const telemetryEvents = spansToExport.map(span =>
        this.spanToTelemetryEvent(span)
      );

      await fetch(this.endpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Finault-Org-ID': this.orgId,
          'X-Finault-API-Key': this.apiKey
        },
        body: JSON.stringify({
          events: telemetryEvents,
          timestamp: new Date().toISOString()
        })
      });

      console.log(`Exported ${telemetryEvents.length} telemetry events`);
    } catch (err) {
      console.error('Failed to flush spans:', err);
      // Re-add spans to buffer for retry
      this.buffer.unshift(...spansToExport);
    }

    // Schedule next flush if buffer still has items
    if (this.buffer.length > 0) {
      this.flushTimer = setTimeout(() => this.flush(), this.flushInterval);
    }
  }

  /**
   * Convert span to Finault telemetry event
   */
  private spanToTelemetryEvent(span: ReadableSpan) {
    const attributes = span.attributes || {};

    return {
      trace_id: span.spanContext().traceId,
      span_id: span.spanContext().spanId,
      parent_span_id: span.parentSpanId,
      name: span.name,
      start_time: span.startTime[0] * 1000 + span.startTime[1] / 1e6, // Convert to ms
      end_time: span.endTime[0] * 1000 + span.endTime[1] / 1e6,
      duration_ms: (span.endTime[0] - span.startTime[0]) * 1000 +
                   (span.endTime[1] - span.startTime[1]) / 1e6,

      // Finault-specific attributes
      finault: {
        cost_usd: attributes['finault.cost'] as number,
        margin_usd: attributes['finault.margin'] as number,
        margin_percent: attributes['finault.margin_percent'] as number,
        seal_id: attributes['finault.seal_id'] as string,
        seal_url: attributes['finault.seal_url'] as string,
        model: attributes['finault.model'] as string,
        provider: attributes['finault.provider'] as string,
        tokens_in: attributes['finault.tokens_in'] as number,
        tokens_out: attributes['finault.tokens_out'] as number,
        cache_hit: attributes['finault.cache_hit'] as boolean,
        cost_method: attributes['finault.cost_method'] as string
      },

      // Standard attributes
      attributes: {
        'http.method': attributes['http.method'],
        'http.url': attributes['http.url'],
        'http.status_code': attributes['http.status_code'],
        'http.host': attributes['http.host'],
        'http.scheme': attributes['http.scheme'],
        'component': attributes['component']
      },

      // Status
      status: {
        code: span.status?.code,
        message: span.status?.message
      },

      // Events (exceptions, etc)
      events: span.events.map(event => ({
        name: event.name,
        timestamp: event.time[0] * 1000 + event.time[1] / 1e6,
        attributes: event.attributes
      }))
    };
  }
}

/**
 * Initialize Finault OTel exporter
 */
export function initFinaultOtel(options: {
  orgId: string;
  apiKey: string;
  endpoint?: string;
  autoInstrument?: boolean;
}): FinaultOtelExporter {
  const exporter = new FinaultOtelExporter({
    orgId: options.orgId,
    apiKey: options.apiKey,
    endpoint: options.endpoint
  });

  return exporter;
}

export default FinaultOtelExporter;
